File: 0012-integer-to-roman/0012-integer-to-roman.cpp
class Solution {
public:
    string intToRoman(int num) {
        unordered_map<int,char> dict = {
            {1,'I'},
            {5,'V'},
            {10,'X'},
            {50,'L'},
            {100,'C'},
            {500,'D'},
            {1000,'M'}
        };
        string n=to_string(num);
        string res="";
        for(int i=0;i<n.length();i++){
            int val = n[i]-'0';
            int units = pow(10,n.length()-i-1);
            if(val<4){
                res+=string(val,dict[units]);
            }
            else if(val>=5 && val<9){
                int mod = val%5;
                res+=string(1,dict[5*units])+string(mod,dict[units]);
            }
            else if(val==4){
                res+=string(1,dict[units]) + string(1,dict[5*units]);
            }
            else if(val==9){
                res+=string(1,dict[units]) + string(1,dict[10*units]);
            }
        }
        return res;
    }
};

File: 0012-integer-to-roman/0012-integer-to-roman.py
class Solution:
    def intToRoman(self, num: int) -> str:
        romans = {1000: "M", 500: "D", 100: "C", 50: "L", 10: "X", 5: "V", 1: "I"}
        ans = ""
        tens:int = 1 
        while num: 
            curr=num%10 
            if curr==4 or curr==9: 
                val = romans[tens]+romans[int((curr+1)*tens)]
            elif curr<=3: 
                val = romans[tens]*curr 
            else: 
                rem = curr%5 
                val = romans[int(5*tens)] + rem*romans[tens]
            tens*=10
            tens=int(tens)
            ans = val+ans
            num=num//10
        return ans
